core/src/main/java/org/aerogear/mobile/core/http/HttpResponse.java

```diff
@@ -18,6 +20,27 @@
      */
     HttpResponse onComplete(Runnable runnable);
 
+    /**
+     * Creates a callback to be called when the request has failed.
+     *
+     * This will be triggered when the request could not be sent or no response can be
+     * received.
+     *
+     * @param runnable a callback method
+     * @return the instance of the HttpResponse for API chaining
+     */
+    HttpResponse onError(Runnable runnable);
+
+    /**
+     * Creates a callback to be called when the request has succeeded.
+     *
+     * This will be triggered when the request succeeded and received a non-error status code.
+     *
+     * @param runnable a callback method
+     * @return the instance of the HttpResponse for API chaining
+     */
+    HttpResponse onSuccess(Runnable runnable);
```

Comment:
doesn't this pretty much make the `onComplete` redundant ? 

I'd assume it's no a "long-running" response, right ?

Reply:
@matzew it's like `try`, `catch`, `finally`. onComplete can always be used to clean things up even when there's an error. And you would use onSuccess only if the request succeeded to do things that rely on the data.

Reply:
@matzew Nopz, it's the same way RxJava handler stuffs

http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/Observer.html

Reply:
@danielpassos actually, nope - `Flowable`/`Observable` are emitting `n` objects - while `Single` just emits one. See the `subscribe(Consumer<? super T> onSuccess, Consumer<? super Throwable> onError)` 

that's why I asked if these are long-running requests, where something like `onNext(ByteBuffer)` would make sense.  


--

Reply:
Right now this API doesn't support that type of request.  When the request gets done things get rolled into memory for better or worse.